### PR TITLE
Specify the key for ES documents, so that subsequent writes will overwrite

### PR DESCRIPTION
**Pull request by colinbm**
## What

* Key the ES document by the S3 object name.
* Add the data (extracted from the object name) to the ES document.

## Why

The S3 object is named by `metric`-`period`-`day`. We can use this to key the ES document, so that if the import is re-ran, it will overwrite the existing entry for the day. Otherwise a new document will be created resulting in duplicate data.

Metric and period are part of the data, but the date is not, and to be useful in ES, it needs to be part of the document, not just in the key.

### CODE REVIEW

File: lib/performance/use_case/send_to_elasticsearch.rb [hunk 1]
```diff
@@ -11,7 +11,10 @@ def initialize(
 
   def execute
     data = @data_fetcher.fetch
-    @elasticsearch_gateway.write data
-    @logger.info "Wrote volumetrics data point '#{data}' to Elasticsearch"
+    date = Date.today.to_s
```
Comment (koetsier) [inline lib/performance/use_case/send_to_elasticsearch.rb:14]: This code isn't called anywhere I believe and can be deleted. The data stamp is best set in the `PerformancePlatform::Gateway::<metric>` classes I think

File: lib/performance/gateway/elasticsearch.rb [hunk 2]
```diff
@@ -5,11 +5,12 @@ def initialize(index)
     @index = index
   end
 
-  def write(data)
+  def write(key, data)
```
Comment (koetsier) [inline lib/performance/gateway/elasticsearch.rb:8]: I wonder if we can get rid of this class altogether. It doesn't seem to do much anymore

Comment (colinbm) [reply lib/performance/gateway/elasticsearch.rb:8]: Unless we're refactoring out the CA structure, this is useful to keep, as it means the `index` can be defined at instantiation in the class, and the use case doesn't need to care about the index.

Comment (colinbm) [reply lib/performance/gateway/elasticsearch.rb:8]: Unless we're refactoring out the CA structure, then it's good to keep this, as it means the `index` can be defined at instantiation in the task, and the use case doesn't care about the index.

File: lib/performance/use_case/send_to_elasticsearch.rb (hunk 1 above)
Comment (colinbm) [reply lib/performance/use_case/send_to_elasticsearch.rb]: Reverted this change - leaving deletion for later.

File: lib/performance/use_case/sync_s3_to_elasticsearch.rb [hunk 3]
```diff
@@ -10,8 +10,11 @@ def initialize(s3_gateway:, elasticsearch_gateway:, logger: Logger.new(STDOUT))
   def execute
     record_count = 0
 
-    @s3_gateway.each do |_key, data|
-      @elasticsearch_gateway.write(data)
+    @s3_gateway.each do |key, data|
```
Comment (koetsier) [inline lib/performance/use_case/sync_s3_to_elasticsearch.rb:13]: maybe we could write a script that adds the 'date' field to all existing records in S3 then we don't  need to do that here. We also need to add a 'date' field to all the `PerformancePlatform::Gateway::<metric>` classes so the right data is uploaded to ES and S3 daily and if we do that then 'new' values in S3 will have a date field and the older ones don't. What do you think?

**Review by koetsier — CHANGES_REQUESTED** [review]

**Review by koetsier — APPROVED** [review]
lgtm

**Pull request merged**